gclient/src/client/backend/gtest.rs

```diff
@@ -0,0 +1,287 @@
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::client::{Backend, Client, Code, Message, Program, TxResult};
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use gear_core::{ids::ProgramId, message::UserStoredMessage};
+use gprimitives::{ActorId, MessageId};
+use gsdk::metadata::runtime_types::gear_common::storage::primitives::Interval;
+use gtest::System;
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    time::{Duration, SystemTime},
+};
+use tokio::{
+    runtime::Builder,
+    sync::{
+        mpsc::{self, UnboundedReceiver, UnboundedSender},
+        Mutex,
+    },
+    task::LocalSet,
+};
+
+/// gear general client gtest backend
+#[derive(Clone)]
+pub struct GTest {
+    tx: UnboundedSender<Request>,
+    resps: Arc<Mutex<HashMap<usize, Response>>>,
+    timeout: Duration,
+    nonce: Arc<AtomicUsize>,
+}
+
+const DEFAULT_TIMEOUT: u64 = 500;
+
+impl GTest {
+    /// New gtest instance with timeout
+    pub fn new(timeout: Duration) -> Result<Self> {
+        let resps = Arc::new(Mutex::new(HashMap::new()));
+        let (tx, rx) = mpsc::unbounded_channel::<Request>();
+        Self::spawn(resps.clone(), rx)?;
+
+        Ok(Self {
+            tx,
+            resps,
+            timeout,
+            nonce: Arc::new(AtomicUsize::new(0)),
+        })
+    }
+
+    /// New general client with `GTest` as backend
+    pub fn client() -> Result<Client<GTest>> {
+        Ok(Client::<GTest> {
+            backend: GTest::new(Duration::from_millis(DEFAULT_TIMEOUT))?,
+        })
+    }
+
+    /// Get gtest result from nonce.
+    async fn resp(&self, nonce: usize) -> Result<Response> {
+        let now = SystemTime::now();
+
+        loop {
+            if now.elapsed()? > self.timeout {
+                return Err(anyhow!("gtest: Transaction timed out!"));
+            }
+
+            if let Some(resp) = self.resps.lock().await.remove(&nonce) {
+                return Ok(resp);
+            }
+        }
+    }
+
+    /// Spawn gtest service
+    fn spawn(
+        resps: Arc<Mutex<HashMap<usize, Response>>>,
+        mut rx: UnboundedReceiver<Request>,
+    ) -> Result<()> {
+        let rt = Builder::new_current_thread().enable_all().build()?;
```

Comment:
To my POV using tokio is overcomplication. What are the benefits? In Sails we do it w/o any use of tokio and it works fine

Reply:
overkilled here since I was solving the sync problem of gtest before, `tokio` in gclient instance stays since it needs to spawn async thread